### ISSUE TITLE
Login to Firebase does not work on Safari 16.1+

### ISSUE THREAD

**Issue by mmmulani** [firebase/firebase-js-sdk#6716]
### Describe your environment

  * Operating System version: Mac OS 12.6
  * Browser version: Safari 16.1
  * Firebase SDK version: 9.10.0
  * Firebase Product: auth

#### Steps to reproduce:
Try logging in to a site using `signInWithRedirect`. Notably it redirects back to the site without an error, and does not log the user in. Checking `getRedirectResult` returns a null user.

This seems like a severe issue as iOS 16.1 will be released next week and presumably this issue will affect all those iOS users.

Notably there seems to be a workaround of using `signInWithPopup`. It seems to work for us, though it is unclear why.

#### Relevant Code:

```javascript
const googleProvider = new GoogleAuthProvider();
signInWithRedirect(getAuth(firebaseApp), googleProvider);
```

**Comment by haydarKarkin** [comment]
Hello, today 16.1 version was released, and this problem still continues. Can you help us with this problem, please?

**Comment by sam-gc** [comment]
Hi folks, we're still diagnosing the particular issues at play, but it appears that `signInWithRedirect` still works as expected if you update your `authDomain` parameter in the Firebase app config to match the URL your site is hosted on (this works if you're on Firebase Hosting).

A full write-up of potential mitigations can be found in https://github.com/firebase/firebase-js-sdk/issues/6443#issuecomment-1187798276

**Comment by mmmulani** [comment]
hey thanks for the update.

just to save anyone else reading this some time, while we have the `authDomain` parameter set up correctly, we do not use Firebase Hosting and thus are still running into this issue. (I tried to follow the mitigations linked but we already have all those configs set up correctly and notably, `signInWithRedirect` works for us on Firefox.)

**Comment by sam-gc** [comment]
Yes, sorry I should have noted that that comment was for Firefox, but we believe it may be a similar underlying mechanism that's causing the user not to be populated.

Mitigation 3 from that list does not use `signInWithRedirect` or `signInWithPopup` and will get around this issue for you.

> You can sign in to the OAuth provider manually, then simply authenticate with Firebase at the end using signInWithCredential(). For example, with Google: https://firebase.google.com/docs/auth/web/google-signin#advanced:-handle-the-sign-in-flow-manually

**Comment by david-wolgemuth** [comment]
We are also experiencing this error.  Using `signInWithRedirect` vs `signInWithPopup` did not help, we are seeing the same issue

I believe this is related to this [prematurely] closed issue: https://github.com/firebase/firebase-js-sdk/issues/6636

**Comment by jzxhuang** [comment]
Can confirm this is happening on 16.1 Safari, including iOS 16.1. The suggested solutions all have flaws: 
1. `signInWithPopup` does not work on mobile web browsers, you must use `signInWithRedirect`
2. Firebase Hosting obviously isn't a universal soln, we aren't using Firebase Hosting
3. Authenticating with the OAuth provider sounds plausible, but this removes one of the huge benefits of using Firebase for authentication in the first place. Lots of eng work, having to implement and maintain OAuth flows manually for each supported provider.

Does it look like there are any other solutions in sight? Can we switch to some cookie-based authentication possibly? As 16.1 rolls out this is going to affect a huge number of users. Thanks! cc @sam-gc @prameshj

**Comment by bvandersloot-mozilla** [comment]
This is likely related to #6443.

**Comment by prameshj** [comment]
> Does it look like there are any other solutions in sight? Can we switch to some cookie-based authentication possibly? As 16.1 rolls out this is going to affect a huge number of users. Thanks!

We are exploring a few options for a longer-term fix. Cookie-based authentication will run into this same limitation since the oauth helper code is in \<project\>.firebaseapp.com domain (or the firebase hosting domain, if specified in the authDomain parameter). If this domain is different from the domain where the app is running, there is a cross-origin storage access to complete sign in, which gets blocked by some browser policies.

Note that the FirebaseUI Web issue(https://github.com/firebase/firebaseui-web/issues/977) is different and presents with a "TypeError: Right side of assignment cannot be destructured" in the console.

**Comment by carlospero** [comment]
Curiously on iOS 16.1, Chrome and Firefox and Brave also suffer the same issue as Safari!

`signInWithPopup` does bypass the problem on all of those browsers too.  So, whatever long-term fix is determined, it hopefully applies across all of iOS.

(Latest versions of Chrome and Firefox and Brave on MacOS still function using `signInWithRedirect`.)

**Comment by galaxyblur** [comment]
> Curiously on iOS 16.1, Chrome and Firefox and Brave also suffer the same issue as Safari!
> 
> `signInWithPopup` does bypass the problem on all of those browsers too. So, whatever long-term fix is determined, it hopefully applies across all of iOS.
> 
> (Latest versions of Chrome and Firefox and Brave on MacOS still function using `signInWithRedirect`.)

There is only one actual browser on iOS: Safari. The other browsers use Safari beneath their own UI/UX.

**Comment by bt-akiko** [comment]
Hello, is anyone meet this issue ? since on my environment, I cannot meet in that means reproduce this phenomenon  iOS 16.1 and safari

**Comment by bvandersloot-mozilla** [comment]
> We are exploring a few options for a longer-term fix. Cookie-based authentication will run into this same limitation since the oauth helper code is in <project>.firebaseapp.com domain (or the firebase hosting domain, if specified in the authDomain parameter). If this domain is different from the domain where the app is running, there is a cross-origin storage access to complete sign in, which gets blocked by some browser policies.

@prameshj:  If you want a privacy engineer at Mozilla to talk to, I can chat about some of these options!

**Comment by deshiknaves** [comment]
Now that Safari 16.1 is released, this is affecting users. Are there any manual work arounds (like implementing the party flow not using the SDK) that we could implement?

Also if there is anything I can help with testing, let me know.

**Comment by deshiknaves** [comment]
Ok. Been doing some digging into this and I can confirm that `signInWithPopup` works exactly as expected as mentioned above. It was mentioned above that this doesn't work on mobile, this is untrue. It opens up another page where you login and it takes you back to the previous window after they have authed. We have a convoluted setup that is a bit more complex that simple popup on a page, and this was not an issue to implement.

I think the one thing that catches people out when implementing `signInWithPopup` is that they initiate it in a way that isn't the direct result of a user interaction. Browsers will block popups unless they are created as a direct result of an action from the user. I.e. put it in the click handler and you won't have an issue.

**Comment by galaxyblur** [comment]
Does `signInWithPopup` work in native app web views? I have to assume it does not, but maybe someone figure it out.

**Comment by Sallaa** [comment]
@galaxyblur we have tried to use `signInWithPopup` with web views for our apps and it did not work for us. However I can also confirm that `signInWithPopup` was a workaround for us for the browser.

**Comment by jzxhuang** [comment]
> Ok. Been doing some digging into this and I can confirm that `signInWithPopup` works exactly as expected as mentioned above. It was mentioned above that this doesn't work on mobile, this is untrue. 

Hmm... I'm pretty sure some mobile web browsers, like Chrome iOS, block popups by default, even when they are the result of an action from the user. Or at least there is a setting to do this. Even if that's not the case, there's still other pitfalls: signInWithPopup will not work with in-app web views, and it can fail for a variety of reasons even on desktop — for example an aggresive adblocker. 

Our strategy is to try `signInWithPopup`, and if that fails with error code `POPUP_BLOCKED`, we fallback to `signInWithRedirect`. `signInWithRedirect` is a failsafe that can be invoked for all sorts of odd reasons/edge cases. 

From [Firebase's Auth docs:](https://firebase.google.com/docs/auth/web/google-signin#popup-mode)

> You can prompt your users to sign in with their Google Accounts either by opening a pop-up window or by redirecting to the sign-in page. The redirect method is preferred on mobile devices.

@prameshj, do you or the Firebase team have an estimate/idea of if there is a fix for this in the short term? Any updates are appreciated — if there is no short-term fix on the horizon, our team needs to seriously consider implementing the OAuth providers ourselves, and only using Firebase SDK to get a FIrebase token (suggestion 3 of https://github.com/firebase/firebase-js-sdk/issues/6443#issuecomment-1187798276). This is obviously a fairly large resource commitment, so we'd like to know if there's any other possible solutions in the works...

**Comment by galaxyblur** [comment]
I 2nd that -- knowing whether a fix in possible in the near term would help a great deal.

**Comment by mesqueeb** [comment]
I also would like to know when a fix is planned. 

If it comes to having to re-implement apple/Google/Twitter auth manually via OAuth, maybe this can be shared among the community here so everyone here can benefit from it.

**Comment by prameshj** [comment]
Hi all, thanks for your patience on this issue.

> Are there any manual work arounds (like implementing the party flow not using the SDK) that we could implement?
   Also if there is anything I can help with testing, let me know.

In addition to the workarounds [listed here](https://github.com/firebase/firebase-js-sdk/issues/6443#issuecomment-1187798276), there is the option to self-host the signin helper code.

Here is a flowchart to determine the most suitable mitigation depending on the setup:

![signInWithRedirect mitigation drawio](https://user-images.githubusercontent.com/35932340/199833760-f60b4046-e6a2-4914-bcd6-52a8190f0630.png)


Self-Hosting the widgets:

Note: This approach will not work for Apple sign in and SAML. Please use a different mitigation for those. 

The `signInWithRedirect` flow starts by redirecting from your app domain to the
domain specified in the `authDomain` parameter in firebase config
("<project>.firebaseapp.com" by default). `authDomain` hosts the sign-in helper
code that redirects to the Identity Provider, which, on success, redirects back
to the app domain.

When authentication flow returns to your app domain, the browser storage of the sign-in helper
domain is accessed. Hosting the helper code in your app domain eliminates the
cross-origin storage access, which otherwise gets blocked by browsers.

Hosting the helper code has the following steps:

1. Download the files to host from the `<project>.firebaseapp.com` location by
executing the following commands:

    ```
    mkdir signin_helpers/ && cd signin_helpers
    wget https://<project>.firebaseapp.com/__/auth/handler
    wget https://<project>.firebaseapp.com/__/auth/handler.js
    wget https://<project>.firebaseapp.com/__/auth/experiment.js
    wget https://<project>.firebaseapp.com/__/auth/iframe.js
    ```

2) Host the above files under your app domain. Ensure that your web server can
respond to `https://<app domain>/__/auth/<filename>`.

 Here is a sample server implementation that downloads and hosts the files - https://go.dev/play/p/OY_MbRC9XZh


3) Follow the steps in[ this comment(option 2)](https://github.com/firebase/firebase-js-sdk/issues/6443#issuecomment-1187798276) to update authorized redirect_uri and your authDomain.
Once you redeploy your app, the cross-origin storage access should no longer happen.

**Comment by prameshj** [comment]
We will document these mitigation options in the official docs and drop a link here soon. We are investigating options for a more permanent solution, however, we do not have an ETA right now.

**Comment by samsargent** [comment]
> Download the files to host from the <project>.firebaseapp.com location by
executing the following commands:

Looks like there is a missing `iframe` html file from this list as the redirect comes back to `/__/auth/iframe?apiKey=...`

Also FYI - `experiment.js` should be `experiments.js`

```
mkdir signin_helpers/ && cd signin_helpers
wget https://<project>.firebaseapp.com/__/auth/handler
wget https://<project>.firebaseapp.com/__/auth/handler.js
wget https://<project>.firebaseapp.com/__/auth/experiments.js
wget https://<project>.firebaseapp.com/__/auth/iframe
wget https://<project>.firebaseapp.com/__/auth/iframe.js
```

**Comment by prameshj** [comment]
Thanks! fixed.

**Comment by lincolnthree** [comment]
> Download the files to host from the <project>.firebaseapp.com location by
executing the following commands...

Out of curiosity, if we are self-hosting. I wonder if setting up a webserver proxy to these files would be a slightly more long-term/resilient option, rather than downloading and hosting the files directly (and risking them going out of date.)

E.g.: Forwarding `https://customdomain.com/__auth/` -- to --> `https://<project>.firebaseapp.com/__/auth`

If so, additionally, I wonder if this is something that might be added as an optional firebase configuration anyway (thus removing the need for the `<project>.firebaseapp.com` domain redirect entirely?

**Comment by sww314** [comment]
@prameshj
Is there a work around for SAML? This is the flow we care about. 
As others have mentioned the `signInWithPopup` does not work on mobile. Native app webviews typically block the popups.

**Comment by prameshj** [comment]
The list of mitigations have been published at https://firebase.google.com/docs/auth/web/third-party-storage-mitigation

> Is there a work around for SAML? This is the flow we care about.

@sww314 if you are not on Firebase Hosting, then https://firebase.google.com/docs/auth/web/third-party-storage-mitigation#mitigation_3_proxy_auth_requests_to_firebaseappcom should work.

**Comment by omrikiei** [comment]
@prameshj we have many customers that use their own SSO providers with our stack of GCP's identity toolkit and changing the redirect domain is not really a valid option for us since we will need all the customers to do so as well, is there an ETA for supporting this issue without a mitigation?

**Comment by kostyay** [comment]
Is there any reason there is no "Cookie" storage? Cookie can work across domains (same sub domains.. we have app.xxx.com and accounts.xxx.com). 
Are there any security reasons not to implement such solution?
Perhaps implementing it would solve the problem?

**Comment by sww314** [comment]
This solution https://firebase.google.com/docs/auth/web/third-party-storage-mitigation#mitigation_3_proxy_auth_requests_to_firebaseappcom
Worked for SAML. 
However it requires our customers update the `Reply URL`. 
In addition for the `Google Provider`, we have to update the `Authorized Redirect URIs`

**Comment by vixeven** [comment]
If you are using Next.js, you can solve this by adding following [rewrite](https://nextjs.org/docs/api-reference/next.config.js/rewrites) to your `next.config.js` file:
```js
/** @type {import('next').NextConfig[]} */
module.exports = {
    rewrites: [
      {
        source: "/__/auth/:path*",
        destination: `https://${firebaseAuthDomain}/__/auth/:path*`,        
      }
    ]
}
```

Note: Don't forget to update `authDomain` in the firebase config and add the new redirect URI in Google Cloud Console.

**Comment by taishi55** [comment]
Since no one talks about the error message causing the issue of `signInWithRedirect `, I will address it here (also mentioned #6831 ). 
`Unrecognized Content-Security-Policy directive 'require-trusted-types-for'.`

**Comment by KyleFoleyImajion** [comment]
If anyone could offer a suggestion for SAML + GAE hosting, that would be fantastic.

**Comment by ffissore** [comment]
We are on GAE, and we fixed it.
What we had:
* our app is served by domain `app.example.com`, served by google app engine
* firebase SDK was configured to work on `auth.example.com`, served by firebase hosting

We fixed it essentially by using this variant of [mitigation #3](https://firebase.google.com/docs/auth/web/third-party-storage-mitigation#mitigation_3_proxy_auth_requests_to_firebaseappcom):
* we changed firebase domain to `app.example.com`
* we deployed a GAE nodejs app that acts as a reverse proxy to `project-id.firebaseapp.com`
* in the main GAE app `dispatch.yaml` to mapped URL `"*/__/auth/*"` to the GAE reverse proxy above
* our app intercepts all taps on URLs `app.example.com` and opens them in the app, which screws up the sign with redirect flow, but in iOS you can add exceptions, so in file `apple-app-site-association` we added the exception `"NOT /__/auth/*"`

This way, all interactions with firebase are now happening from the domain `app.example.com` and everything works as before.

In android instead, since you cannot add exceptions, we are still using `app.example.com` and `auth.example.com`

**Comment by WestonThayer** [comment]
I just wanted to call out that the issue noted by @sww314 https://github.com/firebase/firebase-js-sdk/issues/6716#issuecomment-1320593233 "our customers update the `Reply URL`." is pretty painful when you have a significant number of existing customers who you've integrated your site using SAML/OIDC with (taking the common pattern of having each customer create a custom app in their IdP as opposed to publishing an app in Okta/Auth0/Google Workspace/etc). Unless you take extra steps, all of them would need to update `Reply URL` in lockstep, exactly as your site deploys mitigation `#3`, otherwise their users will be locked out. That's just not a realistic path.

We ended up calling `.deleteApp()` and re-initializing with the proper `authDomain` just before `.signInWithRedirect()` and `.getRedirectResult()`, depending on the customer who was trying to sign in, so that we'd redirect to the same `Reply URL` in their IdP, and can migrate each customer individually.

This was a pain and could have been avoided if Firebase provided us some control over `*.firebaseapp.com/__/auth/*`. For example, configuring it to redirect/pass it's data on to `oursite.com/__/auth/*` would seamlessly fix Safari for all our customers, without the need to manually migrate each one to the new `Reply URL`. I'm disappointed this this sort of control isn't available to us as having an existing customer base to migrate seems like a common scenario.

**Comment by OrhanTozan** [comment]
@mmmulani request to remove "beta" from the issue title, since Safari 16.1 isn't in beta (so that the impact of this bug is better indicated).

**Comment by mmmulani** [comment]
> @mmmulani request to remove "beta" from the issue title, since Safari 16.1 isn't in beta anymore (so that the impact of this bug is better indicated).

good point, thanks

**Comment by OrhanTozan** [comment]
Still happens on Safari 16.2 AFAIK.

**Comment by klon** [comment]
@prameshj Is there a more long term solution being worked on or do we need to settle for the mitigations described above? It is currently very painful to switch `authDomain` when you have several customers relying on SAML SSO with another Reply URL configured.

**Comment by prameshj** [comment]
Hi folks, thanks for your patience on this issue.

Keeping the auth helpers on the same domain as the main app is the best way to future-proof the signInWithRedirect flow. As a result, the mitigations in the [doc](https://firebase.google.com/docs/auth/web/redirect-best-practices) are best-practices for using signInWithRedirect, going forward. We will update the signInWithRedirect docs to reflect this as well.

The popup flow works differently since the main app window is always present and hence the necessary information can be passed along without relying on web storage. All other mitigations aim to keep the auth helper and the main app in the same domain.

Hence, changing the value of `authDomain` is necessary, unfortunately.

**Comment by AlexanderVishnevsky** [comment]
✅ For those who faced with this problem in latest Safari's (16+) there is workaround:
 
 The idea is to proxy your Sign-in requests:
1. If you use something like Nginx, just proxy your request like [shown in the firebase docs](https://firebase.google.com/docs/auth/web/redirect-best-practices#proxy-requests) 
2. Sign-in to your [Google Cloud Console > APIs > Credentials](https://console.cloud.google.com/apis/credentials) 
    - Add your domain to authorised domains 
    - Add redirect URLs as `https://<your domain name>/__/auth/handler`

**Comment by Pickachu** [comment]
As an extra information to the discussion: making the mitigation # 1 (https://github.com/firebase/firebase-js-sdk/issues/6443#issuecomment-1187798276) is not working across sub-domains for me on firebase hosting with custom domains for application and one for authentication. The mitigation must be made for each sub-domain.

Overall conditions on my situation:

- must use `signInWithRedirect` because popup does not work well on iOS devices
- use authentication.mydomain.com as the `authDomain` firebase option in all sub-domains
- have several subdomains (sub-domain-1.mydomain.com, sub-domain-2.mydomain.com, sub-domain-3.mydomain.com) hosting applications on firebase hosting (all implemented on same codebase)


The fix for me was basically doing the mitigation # 1 for each sub-domain:

- For each (sub-domain-1.mydomain.com, sub-domain-2.mydomain.com, sub-domain-3.mydomain.com, etc.) add an authorized redirect URIs
- Keep all sub-domains within firebase hosting


This also adds the overhead to update Authorized redirect URIs for each new sub-domain (essentially a new application), but we add new ones rarely. Eventually will probably go with mitigation # 3 and implement communication between subdomains if needed.

Additionally I could not find any HTTP header or privacy sandbox way of allowing the firebase authentication clients to share information between subdomains.

**Comment by mesqueeb** [comment]
If anyone can share a guide on the steps to take with Vercel hosting, that'd be awesome.
(we have a vite based SPA hosted on Vercel hosting)

would need to know
- What `__/auth` files do I need to put in the repository
- Where in the repository do they need to be placed
- What Vercel side config is needed when the project is uploaded to the Vercel servers so that these files are not part of the SPA but accessible directly

thanks!

**Comment by nwaughachukwuma** [comment]
> ✅ For those who faced with this problem in latest Safari's (16+) there is workaround:
> 
> The idea is to proxy your Sign-in requests:
> 
> 1. If you use something like Nginx, just proxy your request like [shown in the firebase docs](https://firebase.google.com/docs/auth/web/redirect-best-practices#proxy-requests)
> 2. Sign-in to your [Google Cloud Console > APIs > Credentials](https://console.cloud.google.com/apis/credentials)
>    
>    * Add your domain to authorised domains
>    * Add redirect URLs as `https://<your domain name>/__/auth/handler`

Does this work with appengine preview URLs?

**Comment by ingprera** [comment]
**For the following scenario:**
Windows server (IIS) (on-premise)
singInWithRedirect()

I've solved using [mitigation #3](https://firebase.google.com/docs/auth/web/redirect-best-practices#proxy-requests)  

**I've created a reverse proxy with following tools (extensions for IIS)**
[ARR](https://learn.microsoft.com/en-us/iis/extensions/url-rewrite-module/reverse-proxy-with-url-rewrite-v2-and-application-request-routing)
[ URL rewrite](https://learn.microsoft.com/en-us/iis/extensions/url-rewrite-module/creating-rewrite-rules-for-the-url-rewrite-module)

**And I've used the rules in this post (thanks for help):** 

> For those of you using **Azure App Service** on a Windows server to host your app and have this issue using signInWithRedirect() with mobile browsers, here are the steps to resolve using [mitigation #3](https://firebase.google.com/docs/auth/web/redirect-best-practices#proxy-requests).
> 
> 1. In your app hosted at https://yourmainappdomain.com, in the web.config file located inside the wwwroot folder, add the related rewrite rules in the rules section:
> 
> ```
> <rule name="FirebaseAuthRewrite" stopProcessing="true">
>           <match url="^__/auth/handler.js" />
>           <action type="Rewrite" url="https://<yourproject>.firebaseapp.com/__/auth/handler.js" appendQueryString="true" logRewrittenUrl="false" />
>  </rule>
>         <rule name="FirebaseAuthRewriteExperimentsJs" stopProcessing="true">
>           <match url="^__/auth/experiments.js" />
>           <action type="Rewrite" url="https://<yourproject>.firebaseapp.com/__/auth/experiments.js" appendQueryString="true" logRewrittenUrl="false" />
> </rule>
> <rule name="FirebaseAuthRewriteIframeJs" stopProcessing="true">
>           <match url="^__/auth/iframe.js" />
>           <action type="Rewrite" url="https://<yourproject>.firebaseapp.com.com/__/auth/iframe.js" appendQueryString="true" logRewrittenUrl="false" />
> </rule>
> <rule name="FirebaseAuthRewriteHandler" stopProcessing="true">
>           <match url="^__/auth/handler" />
>           <action type="Rewrite" url="https://<yourproject>.firebaseapp.com/__/auth/handler" appendQueryString="true" logRewrittenUrl="false" />
> </rule>
> <rule name="FirebaseAuthRewriteIframe" stopProcessing="true">
>           <match url="^__/auth/iframe" />
>           <action type="Rewrite" url="https://<yourproject>.firebaseapp.com/__/auth/iframe" appendQueryString="true" logRewrittenUrl="false" />
> </rule>
> ```
> 
> You can use a wildcard instead so you have just one rule - but I wanted to list them explicitly, so it is clearer what files are involved. In our case, we are using a Node app so there are additional rules that capture any URLs and send to Node. If this is the case, then you must use stopProcessing="true" as in the above example. If not, you can set that value to false which is default.
> 
> 2. For Azure, there is one more step in order to enable Application Request Routing (ARR) to handle the proxy. Simply create a file named applicationHost.xdt and add to the SAME level as wwwroot, deployments, etc. (not inside wwwroot).
>    As alternative to FTP, in the Azure console, go to your app service, click Development Tools -> Advanced Tools and choose Debug Console and either CMD or Powershell. This will give you the file explorer where you can create and edit both the web.config and the new applicationHost.xdt at the respective levels. Create and edit the applicationHost.xdt to contain the following:
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <configuration xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
>   <system.webServer>
>     <proxy xdt:Transform="InsertIfMissing" enabled="true" preserveHostHeader="false" reverseRewriteHostInResponseHeaders="false" />
>   </system.webServer>
> </configuration>
> ```
> 
> 3. After both the web.config and applicationHost.xdt are configured, you can stop and start your app service. Any changes to either file require a restart of the app service to take effect.
> 4. Finally, change the authDomain in your Firebase config to yourmainappdomain.com to match where your app is hosted. Now  signInWithRedirect() will work on mobile browsers since no third-party storage is used.

**Comment by ishowta** [comment]
Is there any way to make it work in a localhost environment while keeping `signInWithRedirect`?

**Comment by WestonThayer** [comment]
To test locally, you need: some way to publicly expose your localhost (like port forwarding on your router, ngrok); and to add that public IP (or ngrok domain) to the list of authorized domains in GCP Identity Platform or Firebase Console.

**Comment by katopz** [comment]
For me, both Safari and Chrome `109.0.5414.83` on iPad `16.2` have this problem.

**Comment by fjuren** [comment]
Me as well. Safari & Chrome. `signInWithPopup` works fine, `signInWithRedirect` returns user as null and doesn't progress past original login page.

**Comment by OrhanTozan** [comment]
Still happens on Safari 16.3 AFAIK.

**Comment by treeder** [comment]
If firebase were to have a little hosted page that we could set a custom domain on, say `auth.mydomain.com`, to complete this auth process, then redirect back to the real domain we are using at `mydomain.com`, would something like that work? Rather than having us have to pass things through firebase hosting (option 1) or proxy requests (option 3).

**Comment by rshekhtm** [comment]
The Firebase documentation is not quite correct. The correct path to proxy is /__/* and not /__/auth/*. Turns out URLs in email link messages also call out to /__/firebase/init.json and would not sign in properly unless the entire /__/ is proxied.

**Comment by fjuren** [comment]
> If firebase were to have a little hosted page that we could set a custom domain on, say `auth.mydomain.com`, to complete this auth process, then redirect back to the real domain we are using at `mydomain.com`, would something like that work? Then the cookie could be used since they share the same SLD. 
> 
> 
> 
> Rather than having us have to pass things through firebase hosting (option 1) or proxy requests (option 3).  
> 
> 

I'm using firebase hosting but still run into this issue. Have others managed to get it to work while using firebase hosting?

**Comment by JeffersonSchuler** [comment]
We are not using firebase hosting, our app is single page Vue JS based, run locally.  I don't think any of the work arounds are for this kind of situation.

**Comment by prameshj** [comment]
> If firebase were to have a little hosted page that we could set a custom domain on, say `auth.mydomain.com`, to complete this auth process, then redirect back to the real domain we are using at `mydomain.com`, would something like that work?

I think the closest to this is https://firebase.google.com/docs/auth/web/redirect-best-practices#self-host-helper-code

**Comment by prameshj** [comment]
> The Firebase documentation is not quite correct. The correct path to proxy is `/__/*` and not `/__/auth/*`. Turns out URLs in email link messages also call out to `/__/firebase/init.json` and would not sign in properly unless the entire `/__/` is proxied.

The password reset, email link sign in do not use the authDomain in the app config to construct the links. The SDK methods call the rest API -  https://cloud.google.com/identity-platform/docs/reference/rest/v1/projects.accounts/sendOobCode and the backend picks the authDomain (which will be <project>.firebaseapp.com) to construct the link. The proxying + using custom authDomain on the client side should not affect the above flows.

**Comment by hippyhorse** [comment]
## Question 1: Does anyone know what, specifically, has changed with the release of Safari 16.1?

I'd really like to help contribute ideas on fixing, but I'm struggling at the first hurdle to get a clear understanding of what, specifically, has changed in Safari to suddenly make this an issue.

It's suggested in the thread that this may be a cross-origin storage access issue, but looking through the [release notes for Safari 16.1](https://developer.apple.com/documentation/safari-release-notes/safari-16_1-release-notes) there is no explicit mention of any changes to how cross-origin storage access is handled.

Likewise, the [MDN article](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#cross-origin_data_storage_access) on the topic doesnt shed any additional light on what might be different between Safari 16.1, Chrome and others.

## Question 2: Is any long-term solution likely to be forth-coming in the near future?

For example, if the only thing missing here is the addition of some additional cross-origin policy response headers. Then maybe we could extend the 'proxying' workaround, with the addition of those missing headers, so that things start working in Safari again?

Some in this thead have also suggested using Cookies for storage?

As mentioned in question 1 above, with no concrete information on the specific change that Apple has suddenly made, its really hard to think up solutions - so if anyone *does* know, that would be really useful! 

## Our current setup involves a domain & subdomain. We need a SAML solution that doesnt involve customer-side changes

* Our app is a single-page app hosted on our root domain: `example.com`
* We use Google Identity Platform (firebase), and have configured a sub-domain to point to Firebase IP addresses: `auth.example.com`
* Back on `example.com` we load in the Firebase JS SDK and initialize it with `authDomain: 'auth.example.com'`
* All our SAML customers have the SSO URL set to `https://auth.example.com/__/auth/handler`

It is not remotely feasible to go to every single one of our SAML customers and tell them all to migrate before we can fix anything - or at least, if we *had* to go that far, then we'd be migrating away from Firebase at the same time anyway to avoid issues in future.

**Comment by hippyhorse** [comment]
I created a little test rig that reproduces the cause (though I'm still not 100% sure where Safari introduced that change in their release notes)...

## Usage steps

1. Create the file structure below with the two `index.html` files.
2. Start two local web servers (in the same, root folder - using ports `8080` and `8081` - to represent a cross-origin scenario). i.e. `python -m http.server 8080` and `python -m http.server 8080`
3. Now open **a fresh incognito/private browsing window** (note: this will only work once per session)
4. Navigate to the parent (`https://localhost:8080`) and note the console output
5. Now click 'navigate to the iframe itself' and note the console output.

## Observations

See the output variation below - this shows that Safari is entirely isolating iframes when they are loaded from another domain - no local storage retention, no session storage retention, and even no cookie retention. 

### Chrome/Firefox

Initial console output:

```
Local storage is empty
Session storage is empty
Cookie is empty
Updated storage.
```

Output after clicking 'navigate to the iframe itself':

```
Local storage item is set!
Session storage item is set!
Cookie is set!
Updated storage.
```

## Safari 16.1

Initial console output is the same as Chrome/Firefox, but after clicking 'navigate to the iframe itself':

```
Local storage is empty
Session storage is empty
Cookie is empty
Updated storage.
```

## File structure

```
test/
├── iframe/
│   └── index.html
└── index.html
```

### File: `iframe/index.html`

```html
<h1>iframe</h1>

<script>
    console.log(localStorage.getItem('x') ?
        'Local storage item is set!' :
        'Local storage is empty');

    console.log(sessionStorage.getItem('x') ?
        'Session storage item is set!' :
        'Session storage is empty');

    console.log(document.cookie ?
        'Cookie is set!' :
        'Cookie is empty');

    localStorage.setItem('x', '1');
    sessionStorage.setItem('x', '1');
    document.cookie = "x=1; SameSite=None; Secure";

    console.log('Updated storage.');
</script>
```

### File: `index.html`

```html
<h1>Parent document</h1>
<iframe id="iframe" src="http://localhost:8081/iframe/" />
<a href="http://localhost:8081/iframe/">Navigate to the iframe itself</a>
```

**Comment by WestonThayer** [comment]
@hippyhorse Q1: https://github.com/firebase/firebase-js-sdk/issues/6716#issuecomment-1344831136. Relevant https://github.com/cookie-status/cookie-status-dev/issues/44

**Comment by WestonThayer** [comment]
@hippyhorse you can do a gradual migration for your SAML customers using the technique I describe here https://github.com/firebase/firebase-js-sdk/issues/6716#issuecomment-1343600339

If the Firebase team were to give us more control over their redirect URL as I suggest, no customer migration would be necessary.

**Comment by hippyhorse** [comment]
Thanks @WestonThayer coincidentally we were just coming to a similar conclusion that "if we could just dynamically change the redirectURL" things would be much easier to migrate. The technique you've linked to sounds like what we're going to need to do.

Our take away from reading everything, especially the [Webkit team's response](https://bugs.webkit.org/show_bug.cgi?id=247565), is that this isn't going to go away, and moving to a single domain is likely our only option.

**Comment by itsTusharHere** [comment]
I have changed the auth domain in my firebase config and followed all steps related to firebase hosting, still i am able to sign in fine on windows and android machines but as soon as I switch to iOS device all oAuth providers in my case Google, Facebook and Microsoft do not to work anymore. When user clicks on sign in it takes them to their page user does all the steps, it redirects them back to my login page but just re-renders the firebase ui instead of actually signing them in.
The thing to note here is that I have also tried switching to signInWithPopup but with that only google works leaving facebook and microsoft as it is.
`const authorization = firebase.auth();
    
    const uiConfig = {
        callbacks: {
            signInSuccessWithAuthResult : function(authResult, redirectUrl)
            {   
                document.getElementById('loader').style.display = 'block';
                return false;   
            },
            uiShown: function() 
            {
                document.getElementById('loader').style.display = 'none';
            }
        },
        signInOptions: [
         firebase.auth.GoogleAuthProvider.PROVIDER_ID,
         'microsoft.com',
         firebase.auth.FacebookAuthProvider.PROVIDER_ID,
         firebase.auth.EmailAuthProvider.PROVIDER_ID,
         firebase.auth.PhoneAuthProvider.PROVIDER_ID,
        ],
    };

    // Initialize the FirebaseUI Widget using Firebase.
    const ui = new firebaseui.auth.AuthUI(authorization);
    
    // The start method will wait until the DOM is loaded.
    ui.start('#firebaseui-auth-container', uiConfig);`
It's getting really frustrating now, any help is highly appreciated.

**Comment by prameshj** [comment]
> I have changed the auth domain in my firebase config and followed all steps related to firebase hosting, still i am able to sign in fine on windows and android machines but as soon as I switch to iOS device all oAuth providers in my case Google, Facebook and Microsoft do not to work anymore. When user clicks on sign in it takes them to their page user does all the steps, it redirects them back to my login page but just re-renders the firebase ui instead of actually signing them in. The thing to note here is that I have also tried switching to signInWithPopup but with that only google works leaving facebook and microsoft as it is. `const authorization = firebase.auth();
> 
> ```
> const uiConfig = {
>     callbacks: {
>         signInSuccessWithAuthResult : function(authResult, redirectUrl)
>         {   
>             document.getElementById('loader').style.display = 'block';
>             return false;   
>         },
>         uiShown: function() 
>         {
>             document.getElementById('loader').style.display = 'none';
>         }
>     },
>     signInOptions: [
>      firebase.auth.GoogleAuthProvider.PROVIDER_ID,
>      'microsoft.com',
>      firebase.auth.FacebookAuthProvider.PROVIDER_ID,
>      firebase.auth.EmailAuthProvider.PROVIDER_ID,
>      firebase.auth.PhoneAuthProvider.PROVIDER_ID,
>     ],
> };
> 
> // Initialize the FirebaseUI Widget using Firebase.
> const ui = new firebaseui.auth.AuthUI(authorization);
> 
> // The start method will wait until the DOM is loaded.
> ui.start('#firebaseui-auth-container', uiConfig);`
> ```
> 
> It's getting really frustrating now, any help is highly appreciated.

Have you tried using firebaseui 6.0.2? It might be related to https://github.com/firebase/firebaseui-web/issues/977

**Comment by demptd13** [comment]
I have same issue on iPhone XS, iPhone 11 (iOS 16.3.1). But auth is working on Windows and Android, and on iPhone 5s (very old iOS).... I use signInWithRedirect (Google provider)

**Comment by imgyf** [comment]
I'm facing the same issue. Can the Firebase team look at it?

**Comment by demptd13** [comment]
> 

I fix my issue, change signInWithRedirect to signInWithPopup and remove service workers

**Comment by itsTusharHere** [comment]
I finally found a solution, safari does not allow sub domain of a site to access its internal storage, which intern affects all the web browsers running on iOS since they all use webkit. Opting for firebase hosting did not help in my case, the only possible solution I found was to either self-host or to use proxy. The issue lies here with cross domain storage access. In my opinion and my findings [Proxy auth requests to firebaseapp.com](https://firebase.google.com/docs/auth/web/redirect-best-practices#proxy-requests) appears to be the only fruitful way to mitigate this issue.

**Comment by HoshenIshay** [comment]
I have a similar problem
Firebase authentication for an Angular application is not working on iOS version 16
Users who have updated to IOS 16 are unable to log in to the system using a Google account,
whereas users with previous versions can still log in without any issues.

**Comment by SamLoy** [comment]
I was stuck with the same issue. Performing the 'redirect' firebaseui login Facebook login failed on iOS16 both in Safari and through the Facebook WebView. 

As stated above, option 2 here worked for me: https://github.com/firebase/firebase-js-sdk/issues/6443#issuecomment-1187798276

**Comment by dcoleyoung** [comment]
Would it be possible to add an example to the documentation for how to do a reverse proxy in Flask, as we run on app engine on custom domain? for mitigation Option 3: Proxy auth requests to firebaseapp.com

**Comment by prameshj** [comment]
> So, for those stuck with CSP issues, this might help:

> Add the following to script-src policy: 'sha256-4Sry3sxnDNx0KfLx6AAUo8D2CUTWL64/sHuxb6HhCaM=' 'sha256-LhIWOhjLqan1vNuJevmXgr6L87Nor5PqfWzSqz9q74w='

> That is the hash for the two inline scripts as of 2023-03-02

Hi folks, thanks for pointing out this workaround. However, we do not recommend this approach since the script contents might change anytime.

We will add a nonce to the inline script so that can be used as described in https://content-security-policy.com/examples/allow-inline-script/, so it will be a more stable value that can be allowlisted. I will update this thread once we roll this out.

We are working on removing the inline js and will use https://github.com/firebase/firebase-js-sdk/issues/5193 to track this.

**Comment by klemola** [comment]
@prameshj Thank you for the quick response. A nonce is a fair solution, and I'm glad to hear that the inline JS is being removed. This will make the issue much less painful on our end. 

I'll update my message once the nonce is live.

**Comment by CaptainDingle** [comment]
Ever since proxying the iframe requests to our custom domain to solve this issue, we receive the following in the console whenever there is an initial authentication attempt. Everything works in the app but these error keeps appearing. The "bnith" URL request works every request after and only appears one time. The Fingerprint JS will happen every first load only as well. The FingerprintJS error mentions the "customdomain.com/__/auth/iframe?apiKey=...." path.

![2023-03-08_8-08-56](https://user-images.githubusercontent.com/12747005/223721641-5548a8e7-2b50-41a5-9ade-2280162cad87.jpg)

Any ideas regarding this issue is greatly appreciated. Thank you.

**Comment by prameshj** [comment]
> Ever since proxying the iframe requests to our custom domain to solve this issue, we receive the following in the console whenever there is an initial authentication attempt. Everything works in the app but these error keeps appearing. The "bnith" URL request works every request after and only appears one time. The Fingerprint JS will happen every first load only as well. The FingerprintJS error mentions the "customdomain.com/__/auth/iframe?apiKey=...." path.
> 
> ![2023-03-08_8-08-56](https://user-images.githubusercontent.com/12747005/223721641-5548a8e7-2b50-41a5-9ade-2280162cad87.jpg)
> 
> Any ideas regarding this issue is greatly appreciated. Thank you.

Is Fingerprint JS a service like described in https://blog.khophi.co/getting-started-with-fingerprintjs-pro/? I wonder if the error is just saying that  __/auth/iframe code is not hosted in customdomain.com (but it at <project>.firebaseapp.com) and is flagging that as potentially malicious. Is there a way to allowlist this URL path?

**Comment by prameshj** [comment]
> @prameshj Thank you for the quick response. A nonce is a fair solution, and I'm glad to hear that the inline JS is being removed. This will make the issue much less painful on our end.
> 
> I'll update my message to remove the workaround once the nonce is live (to avoid confusion).

Thanks @klemola .. you should now see a nonce around the inline JS in auth/handler and auth/iframe.

```
<script type="text/javascript" nonce="firebase-auth-helper">
var POST_BODY = '{{POST_BODY}}';
fireauth.oauthhelper.widget.initialize();
</script>
```

**Comment by CaptainDingle** [comment]
> Is Fingerprint JS a service like described in https://blog.khophi.co/getting-started-with-fingerprintjs-pro/? I wonder if the error is just saying that __/auth/iframe code is not hosted in customdomain.com (but it at .firebaseapp.com) and is flagging that as potentially malicious. Is there a way to allowlist this URL path?

Thanks for the response @prameshj - Fingerprint JS is not something we are using in our app. It appears to be part of Google's Firebase auth flow via the SDK? It doesn't exist in our project itself.

**Comment by prameshj** [comment]
> > Is Fingerprint JS a service like described in https://blog.khophi.co/getting-started-with-fingerprintjs-pro/? I wonder if the error is just saying that __/auth/iframe code is not hosted in customdomain.com (but it at .firebaseapp.com) and is flagging that as potentially malicious. Is there a way to allowlist this URL path?
> 
> Thanks for the response @prameshj - Fingerprint JS is not something we are using in our app. It appears to be part of Google's Firebase auth flow via the SDK? It doesn't exist in our project itself.

the "getClientIdentity" in the stacktrace isn't coming from the auth sdk or the auth helpers... Do you have more info on where that code is from?

**Comment by logason** [comment]
> Thanks @klemola .. you should now see a nonce around the inline JS in auth/handler and auth/iframe.
> 
> ```
> <script type="text/javascript" nonce="firebase-auth-helper">
> var POST_BODY = '{{POST_BODY}}';
> fireauth.oauthhelper.widget.initialize();
> </script>
> ```

It seems like this change caused our CSP to start blocking Firebase Google login on our site. We already had nonce configured, and injected to every script tag on the page through Nginx. Here's the error when we try to login via Google:

```
Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self' 'nonce-<redacted>"' 'strict-dynamic' https://...<redacted>". Either the 'unsafe-inline' keyword, a hash ('sha256-<redacted>"'), or a nonce ('nonce-...') is required to enable inline execution.
```

When looking at the inline js tags the Firebase SDK pulls in, one of them now has two nonce:
```
<script nonce="<redacted>" type="text/javascript" src="iframe.js"></script>
<script nonce="<redacted>" type="text/javascript" nonce="firebase-auth-helper">
```

Any ideas how to work around this @prameshj? This starting failing few hours ago, sounds like that was around a similar time the `nonce="firebase-auth-helper"` was added.

**Comment by prameshj** [comment]
> > Thanks Matias Klemola .. you should now see a nonce around the inline JS in auth/handler and auth/iframe.
> > ```
> > <script type="text/javascript" nonce="firebase-auth-helper">
> > var POST_BODY = '{{POST_BODY}}';
> > fireauth.oauthhelper.widget.initialize();
> > </script>
> > ```
> 
> It seems like this change caused our CSP to start blocking Firebase Google login on our site. We already had nonce configured, and injected to every script tag on the page through Nginx. Here's the error when we try to login via Google:
> 
> ```
> Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self' 'nonce-<redacted>"' 'strict-dynamic' https://...<redacted>". Either the 'unsafe-inline' keyword, a hash ('sha256-<redacted>"'), or a nonce ('nonce-...') is required to enable inline execution.
> ```
> 
> When looking at the inline js tags the Firebase SDK pulls in, one of them now has two nonce, one from our Nginx and one `nonce="firebase-auth-helper"`:
> 
> ```
> <script nonce="<redacted>" type="text/javascript" src="iframe.js"></script>
> <script nonce="<redacted>" type="text/javascript" nonce="firebase-auth-helper">
> ```
> 
> Any ideas how to work around this @prameshj? This starting failing few hours ago, sounds like that was around a similar time the `nonce="firebase-auth-helper"` was added.

Are you able to use the "firebase-auth-helper" nonce instead of the dynamically generated nginx nonce?  Or can you allowlist "firebase-auth-helper" as an accepted nonce? I believe this would be in the "add-header" section of the nginx config.  Like `add_header Content-Security-Policy "default-src 'self'; script-src 'nonce-firebase-auth-helper';";`

**Comment by logason** [comment]
> Are you able to use the "firebase-auth-helper" nonce instead of the dynamically generated nginx nonce? Or can you allowlist "firebase-auth-helper" as an accepted nonce? I believe this would be in the "add-header" section of the nginx config. Like `add_header Content-Security-Policy "default-src 'self'; script-src 'nonce-firebase-auth-helper';";`

I think the main problem for me right now is that the script tag has two nonce, I assume that won't work. Currently figuring out how I can get Nginx to not inject a nonce for that particular script tag. Right now we're using a very simple `sub_filter` to inject the nonce: `sub_filter '<script' '<script nonce="$cspNonce"';`

**Comment by prameshj** [comment]
> > Are you able to use the "firebase-auth-helper" nonce instead of the dynamically generated nginx nonce? Or can you allowlist "firebase-auth-helper" as an accepted nonce? I believe this would be in the "add-header" section of the nginx config. Like `add_header Content-Security-Policy "default-src 'self'; script-src 'nonce-firebase-auth-helper';";`
> 
> I think the main problem for me right now is that the script tag has two nonce, I assume that won't work. Currently figuring out how I can get Nginx to not inject a nonce for that particular script tag. Right now we're using a very simple `sub_filter` to inject the nonce: `sub_filter '<script' '<script nonce="$cspNonce"';`
> 
> Resulting in two nonce on this element:
> 
> ```
> <script nonce="<redacted>" type="text/javascript" nonce="firebase-auth-helper">
> ```

You could modify the filter to replace `nonce="firebase-auth-helper"` with `nonce="$cspNonce"'`

**Comment by logason** [comment]
> > > Are you able to use the "firebase-auth-helper" nonce instead of the dynamically generated nginx nonce? Or can you allowlist "firebase-auth-helper" as an accepted nonce? I believe this would be in the "add-header" section of the nginx config. Like `add_header Content-Security-Policy "default-src 'self'; script-src 'nonce-firebase-auth-helper';";`
> > 
> > 
> > I think the main problem for me right now is that the script tag has two nonce, I assume that won't work. Currently figuring out how I can get Nginx to not inject a nonce for that particular script tag. Right now we're using a very simple `sub_filter` to inject the nonce: `sub_filter '<script' '<script nonce="$cspNonce"';`
> > Resulting in two nonce on this element:
> > ```
> > <script nonce="<redacted>" type="text/javascript" nonce="firebase-auth-helper">
> > ```
> 
> You could modify the filter to replace `nonce="firebase-auth-helper"` with `nonce="$cspNonce"'

Removing the `nonce="firebase-auth-helper"` resolved the issue for us, with this filter:
```
sub_filter 'nonce="firebase-auth-helper"' '';
```

Why was that nonce being added in the first place? 🤔

**Comment by prameshj** [comment]
> > > > Are you able to use the "firebase-auth-helper" nonce instead of the dynamically generated nginx nonce? Or can you allowlist "firebase-auth-helper" as an accepted nonce? I believe this would be in the "add-header" section of the nginx config. Like `add_header Content-Security-Policy "default-src 'self'; script-src 'nonce-firebase-auth-helper';";`
> > > 
> > > 
> > > I think the main problem for me right now is that the script tag has two nonce, I assume that won't work. Currently figuring out how I can get Nginx to not inject a nonce for that particular script tag. Right now we're using a very simple `sub_filter` to inject the nonce: `sub_filter '<script' '<script nonce="$cspNonce"';`
> > > Resulting in two nonce on this element:
> > > ```
> > > <script nonce="<redacted>" type="text/javascript" nonce="firebase-auth-helper">
> > > ```
> > 
> > 
> > You could modify the filter to replace `nonce="firebase-auth-helper"` with `nonce="$cspNonce"'
> 
> Removing the `nonce="firebase-auth-helper"` resolved the issue for us, with this filter:
> 
> ```
> sub_filter 'nonce="firebase-auth-helper"' '';
> ```
> 
> Why was that nonce being added in the first place? 🤔

See https://github.com/firebase/firebase-js-sdk/issues/6716#issuecomment-1457582094

**Comment by MichelRahme** [comment]
Hi @prameshj,

I followed [Option #4] (https://firebase.google.com/docs/auth/web/redirect-best-practices#self-host-helper-code) but still wasn’t able to actually run the code in the handler file at “myappdomain/__/auth/handler”. It’s always redirecting to my 404 not found page. 

The other options are not feasible for me since my hosting is done on cloudfront and a pop up won’t be optimal for my users. 

I must be missing something. Do you have an idea on what might be happening?

**Comment by prameshj** [comment]
> Hi @prameshj, everyone,
> 
> I followed [Option #4] (https://firebase.google.com/docs/auth/web/redirect-best-practices#self-host-helper-code) but still wasn’t able to actually run the code in the handler file at “myappdomain/__/auth/handler”. It’s always downloading the handler file instead of executing it.
> 
> The other options are not feasible for me since my hosting is done on cloudfront and a pop up won’t be optimal for my users.
> 
> I must be missing something. Do you have an idea on what might be happening?

Can you ensure that calls to GET  “myappdomain/__/auth/handler” are returning html back? This will look similar to 
```
<!DOCTYPE html>
<html>
<head>
<meta name=viewport content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<script type="text/javascript" src="experiments.js"></script>
<script type="text/javascript" src="handler.js"></script>
...
...
```

You can create an HAR file in the browser to ensure the calls to download handler, followed by handler.js, experiment.js are all being issued.

**Comment by MichelRahme** [comment]
Fixed @prameshj

**Comment by CaptainDingle** [comment]
> > > Is Fingerprint JS a service like described in https://blog.khophi.co/getting-started-with-fingerprintjs-pro/? I wonder if the error is just saying that __/auth/iframe code is not hosted in customdomain.com (but it at .firebaseapp.com) and is flagging that as potentially malicious. Is there a way to allowlist this URL path?
> > 
> > 
> > Thanks for the response @prameshj - Fingerprint JS is not something we are using in our app. It appears to be part of Google's Firebase auth flow via the SDK? It doesn't exist in our project itself.
> 
> the "getClientIdentity" in the stacktrace isn't coming from the auth sdk or the auth helpers... Do you have more info on where that code is from?

We found that this actually had nothing to do with Firebase or this open issue at all. The error and code injection was caused by our web application firewall. We just happen to notice it at the time we added the Firebase proxy to work around this open issue and drew an incorrect conclusion. Thank you.

**Comment by prameshj** [comment]
> > > > Is Fingerprint JS a service like described in https://blog.khophi.co/getting-started-with-fingerprintjs-pro/? I wonder if the error is just saying that __/auth/iframe code is not hosted in customdomain.com (but it at .firebaseapp.com) and is flagging that as potentially malicious. Is there a way to allowlist this URL path?
> > > 
> > > 
> > > Thanks for the response @prameshj - Fingerprint JS is not something we are using in our app. It appears to be part of Google's Firebase auth flow via the SDK? It doesn't exist in our project itself.
> > 
> > 
> > the "getClientIdentity" in the stacktrace isn't coming from the auth sdk or the auth helpers... Do you have more info on where that code is from?
> 
> We found that this actually had nothing to do with Firebase or this open issue at all. The error and code injection was caused by our web application firewall. We just happen to notice it at the time we added the Firebase proxy to work around this open issue and drew an incorrect conclusion. Thank you.

Thank you so much for closing the loop here!

**Comment by prameshj** [comment]
Closing this issue since following the best practices at https://firebase.google.com/docs/auth/web/redirect-best-practices fixes this.

**Comment by ahetawal-p** [comment]
I am having similar issues like this comment here: https://github.com/firebase/firebase-js-sdk/issues/6716#issuecomment-1482193328
The `handler` and `iframe` files are being downloaded from browser instead of being rendered. 
I am using nextjs + vercel to deploy my app, and really not sure how do I proceed. ?

**Comment by tycrimm** [comment]
> I am having similar issues like this comment here: [#6716 (comment)](https://github.com/firebase/firebase-js-sdk/issues/6716#issuecomment-1482193328) The `handler` and `iframe` files are being downloaded from browser instead of being rendered. I am using nextjs + vercel to deploy my app, and really not sure how do I proceed. ?

Adding the following rewrite to `vercel.json` and updating the authorized domains/redirect_urls in the [GCP Console](https://console.cloud.google.com/apis/credentials) makes the approach of [proxying requests](https://firebase.google.com/docs/auth/web/redirect-best-practices#proxy-requests) viable: 

```
{
  ...,
  "rewrites": [
    {
      "source": "/__/auth/(.*)",
      "destination": "https://YOUR_APP_ID.firebaseapp.com/__/auth/$1"
    },
  ],
  ...
}

```
